### PR TITLE
core: introduce evenly distributed int64 random in range

### LINKED ISSUE #5075: Implement functions for getting a random value in range
In #4609 it was revealed we have a signed integer overflow in `test/unit/histogram.c` in `gen_rand_value()`. It tries to get a random value in a range. Sometimes it happens that the range length exceeds `INT64_MAX`, and `min - max` expression overflows `int64_t`.
Since this is not the only place, where it is necessary to get a random value in a range, would be good to solve this task once for all places, somewhere in `core/random.c` and `.h`. Without undefined behaviour, and preserving the random generator distribution (it is not ok to use things like `rand() % range`).

### PR DESCRIPTION
**Pull request by PersDep**
Tarantool codebase had at least two functions to generate random
integer in given range and both of them had problems at least with
integer overflow. This patch brings nice functions to generate random
int64_t in given range without overflow while preserving uniform random
generator distribution using unbiased bitmask with rejection method.
It is now possible to use xoshiro256++ PRNG or random bytes as a basis.
Most relevant replacements have been made. Needed tests are introduced.

Closes #5075

### CODE REVIEW

**Comment by PersDep** [comment]
1 LGTM from @Gerold103

**Review by alyapunov — APPROVED** [review]

File: src/lib/core/random.c [hunk 1]
```diff
@@ -97,3 +103,82 @@ random_bytes(char *buf, size_t size)
...
+int64_t
+real_random_in_range(int64_t min, int64_t max)
+{
+	assert(max >= min);
+	uint64_t range = (uint64_t)max - min;
+	uint64_t mask = UINT64_MAX >> bit_clz_u64(range | 1);
+	uint64_t result;
+	do {
+		result = real_random() & mask;
+	} while (result > range);
```
Comment (Korablev77) [inline src/lib/core/random.c:169]: This cycle doesn't seem to be trivial (at least from the first sight). For instance, in swim_scaled_rand()/gen_rand_value() calculation was obvious. Could you please supply it with a comment?

File: src/lib/swim/swim.c [hunk 2]
```diff
@@ -188,25 +189,6 @@ enum {
...
 
-/**
- * Return a random number within given boundaries.
- *
- * Instead of blindly calculating a modulo, scale the random
- * number down the given boundaries to preserve the original
- * distribution. The result belongs to range [start, end].
- */
-static inline int
-swim_scaled_rand(int start, int end)
```
Comment (Korablev77) [inline src/lib/swim/swim.c:199]: I suggest to move usage replacement of new random functions to a separate patch. Up to you

File: src/lib/core/random.c (hunk 1 above)
Comment (Korablev77) [reply src/lib/core/random.c:169]: Okay, to understand how it is supposed to work one should read this article https://www.pcg-random.org/posts/bounded-rands.html

Comment (PersDep) [reply src/lib/core/random.c:169]: Yes, this  article is mentioned in my comment for the issue https://github.com/tarantool/tarantool/issues/5075#issuecomment-701052530
Should I also specify it in comment among the code?

Comment (Korablev77) [reply src/lib/core/random.c:169]: For sure, link to the article and brief explanation (in case link becomes invalid/outdated).

**Comment by Korablev77** [comment]
Also consider using new random functions in SQL. For example: https://github.com/tarantool/tarantool/commit/5a17978664b6c5f3b216e850f8b05d7bfb9ec4e3

**Comment by PersDep** [comment]
> Also consider using new random functions in SQL. For example: [5a17978](https://github.com/tarantool/tarantool/commit/5a17978664b6c5f3b216e850f8b05d7bfb9ec4e3)

Well yes, I considered those but decided to leave it for follow-ups

File: src/lib/core/random.c (hunk 1 above)
Comment (PersDep) [reply src/lib/core/random.c:184]: Added

**Comment by Korablev77** [comment]
Pushed to master, 2.6 and 2.5. Closing PR

**Pull request closed without merging**